### PR TITLE
RD 3761 - pymongo preparations

### PR DESCRIPTION
**Pull request by saartochner**

### CODE REVIEW

File: src/lumigo_tracer/tracer.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    We are adding lumigo's flag to the context, and check it's value in _is_context_already_wrapped.
+    """
+    if len(args) >= 2:
+        with lumigo_safe_execute("wrap context"):
+            setattr(args[1], CONTEXT_WRAPPED_BY_LUMIGO_KEY, True)
+
+
+def _lumigo_tracer(func):
+    @wraps(func)
+    def lambda_wrapper(*args, **kwargs):
```
Comment (doriaviram) [inline src/lumigo_tracer/tracer.py:42]: Check this condition outside the lambda wrapper, and just return `func` if it switched off

File: src/lumigo_tracer/tracer.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    )
+
+
+def _add_prefix_for_each_line(prefix: str, text: str):
+    enhanced_lines = []
+    for line in text.split("\n"):
+        if line and not line.startswith(prefix):
+            line = prefix + " " + line
+        enhanced_lines.append(line)
+    return "\n".join(enhanced_lines)
```
Comment (doriaviram) [inline src/lumigo_tracer/tracer.py:105]: Lets talk, I think all this part is useless

File: src/lumigo_tracer/tracer.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            finally:
+                SpansContainer.get_span().end(ret_val)
+                if Configuration.enhanced_print:
+                    builtins.print = local_print
+                    logging.Formatter.format = local_logging_format
+            return ret_val
+        except Exception:
+            # The case where our wrapping raised an exception
+            if not executed:
+                TimeoutMechanism.stop()
```
Comment (doriaviram) [inline src/lumigo_tracer/tracer.py:77]: Not any way we want to stop the `TimeoutMechanism` ?

**Review by doriaviram — CHANGES_REQUESTED** [review]
Good job !

File: src/lumigo_tracer/event/event_trigger.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+def _parse_streams(event: dict) -> Dict[str, str]:
+    """
+    :return: {"triggeredBy": str, "arn": str}
+    If has messageId, return also: {"messageId": str}
+    """
+    triggered_by = event["Records"][0]["eventSource"].split(":")[1]
+    result = {"triggeredBy": triggered_by}
+    if triggered_by == "s3":
+        result["arn"] = event["Records"][0]["s3"]["bucket"]["arn"]
```
[the remark this answers is not in the document]
Comment (saartochner) [reply src/lumigo_tracer/event/event_trigger.py:152]: I don't want it to be safe :)
If I'm failing - then I want that the trigger will be unknown 🤔

[the remark this answers is not in the document]
Comment (saartochner) [reply src/lumigo_tracer/event/event_trigger.py:152]: the "safe" is above, in `parse_triggered_by`, that fallback to unknown in error

File: src/lumigo_tracer/wrappers/http/sync_http_wrappers.py [hunk 5]
```diff
@@ -36,6 +23,64 @@
...
+
+def add_unparsed_request(parse_params: HttpRequest):
+    """
+    This function handle the case where we got a request the is not fully formatted as we expected,
+    I.e. there isn't '\r\n' in the request data that <i>logically</i> splits the headers from the body.
+
+    In that case, we will consider it as a continuance of the previous request if they got the same url,
+        and we didn't get any answer yet.
+    """
+    last_event = SpansContainer.get_span().get_last_span()
```
[the remark this answers is not in the document]
Comment (saartochner) [reply src/lumigo_tracer/wrappers/http/sync_http_wrappers.py:44]: you're right,
we absolutely assume synchronous executions! (That's why we called it `sync_http_wrapper`)

**Review by doriaviram — APPROVED** [review]

**Pull request merged**